### PR TITLE
streamlink: 3.2.0 -> 5.0.1; pythonPackages.versioningit: init at 2.0.1

### PR DESCRIPTION
**Pull request by DeeUnderscore**
###### Description of changes
This is #171179 plus updates to versioningit and Streamlink, reopened as new PR so all those force pushes don't make it hard to follow.

Changelogs:
* https://github.com/streamlink/streamlink/releases/tag/4.0.0
* https://github.com/streamlink/streamlink/releases/tag/4.0.1
* https://github.com/streamlink/streamlink/releases/tag/4.1.0
* https://github.com/streamlink/streamlink/releases/tag/4.2.0

In 4.0.0, Streamlink switched to the PEP 518 format (pyproject.toml), and additionally introduced a build-time dependency on versioningit (although the PyPI source distribution has the version number substituted in, so we don't need the full Git history). 

versioningit does not build below Python 3.8 due to transitive dependencies in Nixpkgs not building below 3.8, so I'm constraining Streamlink likewise (and omitting a <3.8 dependency).

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by pbsds — APPROVED** [review]

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 180910` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.versioningit</li>
    <li>python39Packages.versioningit</li>
    <li>streamlink</li>
    <li>streamlink-twitch-gui-bin</li>
  </ul>
</details>

File: pkgs/development/python-modules/versioningit/default.nix [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  pname = "versioningit";
+  version = "2.0.0";
+  disabled = pythonOlder "3.8";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-wWkTrzW76WeXuqOgkQ/GCxwF0lrFkd72Ij+oBvmcvlg=";
+  };
+
+  format = "pyproject";
```
Comment (jonringer) [inline pkgs/development/python-modules/versioningit/default.nix:28]:
```suggestion
  disabled = pythonOlder "3.8";
  format = "pyproject";

  src = fetchPypi {
    inherit pname version;
    sha256 = "sha256-wWkTrzW76WeXuqOgkQ/GCxwF0lrFkd72Ij+oBvmcvlg=";
  };

```
most common to punch all of the "high level" configuration toward the top

**Review by jonringer — COMMENTED** [review]
otherwise diff lgtm

**Comment by DeeUnderscore** [comment]
Updated to [4.3.0](https://github.com/streamlink/streamlink/releases/tag/4.3.0) with rebase on master.

**Comment by DeeUnderscore** [comment]
Updated Streamlink to [5.0.0](https://github.com/streamlink/streamlink/releases/tag/5.0.0), versioningit init to 2.0.1 (was 2.0.0), with rebase on master.

**Comment by DeeUnderscore** [comment]
Updated Streamlink to [5.0.1](https://github.com/streamlink/streamlink/releases/tag/5.0.1) with rebase on master

File: pkgs/applications/video/streamlink/default.nix [hunk 2]
```diff
@@ -6,11 +6,13 @@
 
 python3Packages.buildPythonApplication rec {
   pname = "streamlink";
-  version = "3.2.0";
+  version = "5.0.1";
+  disabled = python3Packages.pythonOlder "3.8";
```
Comment (SuperSandro2000) [inline pkgs/applications/video/streamlink/default.nix:10]:
```suggestion
```
that is never the case

File: pkgs/development/python-modules/versioningit/default.nix [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    tomli
+  ] ++ lib.optionals (pythonOlder "3.10") [
+    importlib-metadata
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+    build
+    pydantic
+    pytest-cov
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/versioningit/default.nix:41]:
```suggestion
```
we don't need that

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/versioningit/default.nix [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    # wants to write to the Nix store
+    "test_editable_mode"
+  ];
+
+  preCheck = ''
+    substituteInPlace tox.ini \
+      --replace "--cov=versioningit" "" \
+      --replace "--cov-config=tox.ini" "" \
+      --replace "--no-cov-on-fail" ""
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/versioningit/default.nix:55]:
```suggestion
```

File: pkgs/development/python-modules/versioningit/default.nix [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+buildPythonPackage rec {
+  pname = "versioningit";
+  version = "2.0.1";
+  disabled = pythonOlder "3.8";
+  format = "pyproject";
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-gJfiYNm99nZYW9gTO/e1//rDeox2KWJVtC2Gy1EqsuM=";
+  };
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/versioningit/default.nix:26]:
```suggestion
  };

  postPatch = ''
    substituteInPlace tox.ini \
      --replace "--cov=versioningit" "" \
      --replace "--cov-config=tox.ini" "" \
      --replace "--no-cov-on-fail" ""
  '';
```
